## Move images into folders based on theirt classification
The previous program stored the images in one single folder, this code attempts to classify it and store<br>
it in folders based on it's name<br>
It also analyses the data that exists in order to design algorithms in the future

In [6]:
import os
import shutil

In [7]:
for count, filename in enumerate(os.listdir("agroimages/combined_images")):
    no = count
print(no)

3739


Loops thrugh the combined images folder

In [8]:
number = 30
for count, filename in enumerate(os.listdir("agroimages/combined_images")):
    if count>= number:
        break
    else:
        print(filename)

Maize Kharif Early Vegetative  Stem Borer  No Disease  No Deficiency955.jpg
Red Gram Kharif Early Vegetative  No Insect  No Disease  No Deficiency498.jpg
Cotton Kharif Squaring  Mealybugs  No Disease  No Deficiency921.jpg
Cotton Kharif Early Vegetative  No Insect  No Disease  No Deficiency480.jpg
Maize Kharif Grain Filling  Cob Borer  No Disease  No Deficiency39.jpg
Cotton Kharif Early Vegetative  Mealybugs  No Disease  No Deficiency19.jpg
Cotton Kharif Early Vegetative  No Insect  No Disease  No Deficiency331.jpg
Cotton Kharif Early Vegetative  Mealybugs  No Disease  No Deficiency401.jpg
Cotton Kharif Early Vegetative  No Insect  No Disease  No Deficiency457.jpg
Cotton Kharif Tillering  No Insect  No Disease  No Deficiency398.jpg
Red Gram Kharif Early Vegetative  No Insect  No Disease  No Deficiency329.jpg
Maize Kharif Grain Filling  No Insect  Downy mildew  No Deficiency85.jpg
Black Gram Kharif Early Maturity  No Insect  Others  No Deficiency529.jpg
Maize leafspot228.jpg
Cotton Khari

Two helper functions that we will need

In [9]:
import re
def remove_second_group(m):
    """ Return only group 1 from the match object
        Delete other groups """
    return m.group(1)

In [10]:
#Removes numbers from the back of a string
def remove_ints(m):
    if m[-1].isalpha():
        return m
    else:
        return remove_ints(m[:-1])

In [12]:
number = 10
for count, filename in enumerate(os.listdir("agroimages/combined_images")):
    if count>= number:
        break
    else:
        result = re.sub("(.*)(.{4}$)", remove_second_group, filename) # remove the .jpg from the name
        # now we need to get rid of the numbers at the end
        result = remove_ints(result)
        print(result)

Maize Kharif Early Vegetative  Stem Borer  No Disease  No Deficiency
Red Gram Kharif Early Vegetative  No Insect  No Disease  No Deficiency
Cotton Kharif Squaring  Mealybugs  No Disease  No Deficiency
Cotton Kharif Early Vegetative  No Insect  No Disease  No Deficiency
Maize Kharif Grain Filling  Cob Borer  No Disease  No Deficiency
Cotton Kharif Early Vegetative  Mealybugs  No Disease  No Deficiency
Cotton Kharif Early Vegetative  No Insect  No Disease  No Deficiency
Cotton Kharif Early Vegetative  Mealybugs  No Disease  No Deficiency
Cotton Kharif Early Vegetative  No Insect  No Disease  No Deficiency
Cotton Kharif Tillering  No Insect  No Disease  No Deficiency


In [13]:
number = 5000
Dict = {}
for count, filename in enumerate(os.listdir("agroimages/combined_images")):
    if count>= number:
        break
    else:
        result = re.sub("(.*)(.{4}$)", remove_second_group, filename) # remove the .jpg from the name
        # now we need to get rid of the numbers at the end
        result = remove_ints(result)
        result = result.title() # capitalizes after space to fix errors
        result = result.replace(" ", "")
        if result not in Dict.keys():
            Dict[result] = 1
        else:
            Dict[result] = Dict[result]+1
print(len(Dict.keys()))

167


#### number of images per classification
This number of images per layer is very low as this is real world data and data will be added to it over time 

In [14]:
for x in Dict:
    print (x, Dict[x])

MaizeKharifEarlyVegetativeStemBorerNoDiseaseNoDeficiency 228
RedGramKharifEarlyVegetativeNoInsectNoDiseaseNoDeficiency 501
CottonKharifSquaringMealybugsNoDiseaseNoDeficiency 24
CottonKharifEarlyVegetativeNoInsectNoDiseaseNoDeficiency 519
MaizeKharifGrainFillingCobBorerNoDiseaseNoDeficiency 105
CottonKharifEarlyVegetativeMealybugsNoDiseaseNoDeficiency 164
CottonKharifTilleringNoInsectNoDiseaseNoDeficiency 24
MaizeKharifGrainFillingNoInsectDownyMildewNoDeficiency 24
BlackGramKharifEarlyMaturityNoInsectOthersNoDeficiency 4
MaizeLeafspot 1
MaizeKharifGrainFillingStemBorerNoDiseaseNoDeficiency 77
CottonKharifEarlyVegetativeLeafHopperNoDiseaseNoDeficiency 311
CottonKharifEarlyVegetativeNoInsectCottonLeafCurlVirusNoDeficiency 60
CottonKharifSquaringNoInsectNoDiseaseNoDeficiency 361
RedGramKharifEarlyVegetativeNoInsectSterilityMosaicNoDeficiency 7
CottonKharifSquaringWhiteFliesNoDiseaseNoDeficiency 22
MaizeKharifGrainFillingNoInsectMaydisLeafBlightNoDeficiency 19
RiceKharifEarlyVegetativeNoIns

In [15]:
# lets create folders now 

In [109]:
number = 5000
# Parent Directory path
parent_dir = "agroimages/classified"
Dict = {}
for count, filename in enumerate(os.listdir("agroimages/combined_images")):
    
    result = re.sub("(.*)(.{4}$)", remove_second_group, filename) # remove the .jpg from the name
    # now we need to get rid of the numbers at the end
    result = remove_ints(result)
    result = result.title()
    result = result.replace(" ", "") # removes spaces
    
    if result not in Dict.keys():
        Dict[result] = 1
    else:
        Dict[result] = Dict[result]+1

    #shutil.copy("agroimages/Combined_images/" + filename, path)


In [110]:
for x in Dict.keys():
    directory = x
    path = os.path.join(parent_dir, directory)
    os.makedirs(path, exist_ok=True)    # creates the path

In [108]:
# populate the folders 

In [111]:
number = 5000
# Parent Directory path
parent_dir = "agroimages/classified"
Dict = {}
for count, filename in enumerate(os.listdir("agroimages/combined_images")):
    
    result = re.sub("(.*)(.{4}$)", remove_second_group, filename) # remove the .jpg from the name
    # now we need to get rid of the numbers at the end
    result = remove_ints(result)
    result = result.title()
    result = result.replace(" ", "")
    if result not in Dict.keys():
        Dict[result] = 1
    else:
        Dict[result] = Dict[result]+1
    if "agroimages/Combined_images/" + filename != 'agroimages/Combined_images/.ipynb_checkpoints':
        path = os.path.join(parent_dir, result)
        shutil.copy("agroimages/Combined_images/" + filename, path)

    

In [ ]:
# Remove empty folders?

# Analysis

In [53]:
c =0
num = 0
amt =0
for x in Dict:
        num = num + Dict[x]
amt = c* num
print (num)

3740


In [54]:
c =0
num = 0
for x in Dict:
    if "NoDisease" in x:
        c=c+1
        num = num + Dict[x]

print (num)

3395


In [74]:
c =0
num = 0
for x in Dict:
    if "Deficiency" in x:
        c=c+1
        num = num + Dict[x]

print (num)

3382


In [75]:
c =0
num = 0
for x in Dict:
    if "Insect" in x:
        c=c+1
        num = num + Dict[x]

print (num)

2266


In [26]:
c =0
num = 0
amt =0
for x in Dict:
    if "Maize" in x:
        c=c+1
        num = num + Dict[x]
amt = c* num
print (num)

803


In [27]:
c =0
num = 0
for x in Dict:
    if "RedGram" in x:
        c=c+1
        num = num + Dict[x]

print (num)

566


In [28]:
c =0
num = 0
for x in Dict:
    if "Cotton" in x:
        c=c+1
        num = num + Dict[x]

print (num)

1881


In [29]:
c =0
num = 0
for x in Dict:
    if "BlackGram" in x:
        c=c+1
        num = num + Dict[x]

print (num)

136


In [52]:
c =0
num = 0
for x in Dict:
    if "Rice" in x:
        c=c+1
        num = num + Dict[x]

print (num)

350


## Cotton

In [48]:
c =0
num = 0
for x in Dict:
    if "Cotton" in x and "Disease" not in x:
        print(x, Dict[x])

CottonKharifEarlyVegetativeNoInsectCottonLeafCurlVirusNoDeficiency 60
CottonKharifTilleringLeafMinor 2
CottonKharifTilleringNoInsectCottonLeafCurlVirusNoDeficiency 25
CottonKharifSquaringNoInsectCottonLeafCurlVirusNoDeficiency 21
CottonKharifEarlyVegetativeLeafHopperCottonLeafCurlVirusNoDeficiency 10
CottonKharifFloweringNoInsectCottonLeafCurlVirusNoDeficiency 4
CottonKharifEarlyVegetativeNoInsectAlternariaLeafSpotNoDeficiency 14


In [78]:
c =0
num = 0
for x in Dict:
    if "Cotton" in x and "Insect" not in x:
        print(x, Dict[x])

CottonKharifSquaringMealybugsNoDiseaseNoDeficiency 24
CottonKharifEarlyVegetativeMealybugsNoDiseaseNoDeficiency 164
CottonKharifEarlyVegetativeLeafHopperNoDiseaseNoDeficiency 311
CottonKharifSquaringWhiteFliesNoDiseaseNoDeficiency 22
CottonKharifEarlyVegetativeAphidNoDiseaseNoDeficiency 42
CottonKharifFloweringLeafHopperNoDiseaseNoDeficiency 43
CottonKharifSquaringLeafHopperNoDiseaseNoDeficiency 58
CottonKharifTilleringLeafMinor 2
CottonKharifEarlyVegetativeCutWormNoDiseaseNoDeficiency 6
CottonKharifTilleringLeafHopperNoDiseaseNoDeficiency 18
CottonKharifEarlyVegetativeNoDeficiencyNoDiseaseNoDeficiency 2
CottonKharifEarlyVegetativeLeafHopperCottonLeafCurlVirusNoDeficiency 10
CottonKharifSquaringThripsNoDiseaseNoDeficiency 1
CottonKharifEarlyVegetativeWhiteFliesNoDiseaseNoDeficiency 24
CottonKharifFloweringWhiteFliesNoDiseaseNoDeficiency 6
CottonKharifTilleringWhiteFliesNoDiseaseNoDeficiency 8
CottonKharifBollFormationLeafHopperNoDiseaseNoDeficiency 5
CottonKharifTilleringLeafHopperAphi

In [79]:
c =0
num = 0
for x in Dict:
    if "Insect" in x and "Cotton" in x:
        c=c+1
        num = num + Dict[x]

print (num)

1111


In [87]:
c =0
num = 0
for x in Dict:
    if "Cotton" in x and "CutWorm" in x:
        c=c+1
        num = num + Dict[x]

print (num)

7


In [49]:
c =0
num = 0
for x in Dict:
    if "Cotton" in x and "LeafCurl" in x:
        c=c+1
        num = num + Dict[x]

print (num)

121


In [50]:
c =0
num = 0
for x in Dict:
    if "Cotton" in x and "LeafSpot" in x:
        c=c+1
        num = num + Dict[x]

print (num)

16


## Maize

In [41]:
c =0
num = 0
for x in Dict:
    if "Maize" in x:
        print(x, Dict[x])

MaizeKharifEarlyVegetativeStemBorerNoDiseaseNoDeficiency 228
MaizeKharifGrainFillingCobBorerNoDiseaseNoDeficiency 105
MaizeKharifGrainFillingNoInsectDownyMildewNoDeficiency 24
MaizeLeafspot 1
MaizeKharifGrainFillingStemBorerNoDiseaseNoDeficiency 77
MaizeKharifGrainFillingNoInsectMaydisLeafBlightNoDeficiency 19
MaizeKharifEarlyVegetativeNoInsectNoDiseaseSulphur 48
MaizeKharifGrainFillingNoInsectNoDiseaseNoDeficiency 108
MaizeKharifTilleringStemBorerNoDiseaseSulphur 6
MaizeEarlyKharifLeafSpotLeaf 3
MaizeKharifGrainFillingNoInsectNoDiseaseSulphur 11
MaizeKharifEarlyVegetativeNoInsectNoDiseaseNoDeficiency 51
MaizeEarlyKharifEyeSpot,StemBorerLeaf 3
MaizeKharifEarlyVegetativeStemBorerNoDiseaseSulphur 18
Maize 9
MaizeEarlyKharifLeafBlight,EyeSpotLeaf 1
MaizeKharifEarlyVegetativeCobBorerNoDiseaseNoDeficiency 14
MaizeEarlyKharifEyeSpotLeaf 9
MaizeKharifEarlyVegetativeStemBorerNoInsectNoDiseaseNoDeficiency 2
MaizeEarlyKharifHealthyLeaf 1
MaizeKharifCobFormationSulfurDeficiencyLeaf 1
MaizeKharifG

In [88]:
c =0
num = 0
for x in Dict:
    if "Maize" in x and "Disease" not in x:
        print(x, Dict[x])

MaizeKharifGrainFillingNoInsectDownyMildewNoDeficiency 24
MaizeLeafspot 1
MaizeKharifGrainFillingNoInsectMaydisLeafBlightNoDeficiency 19
MaizeEarlyKharifLeafSpotLeaf 3
MaizeEarlyKharifEyeSpot,StemBorerLeaf 3
Maize 9
MaizeEarlyKharifLeafBlight,EyeSpotLeaf 1
MaizeEarlyKharifEyeSpotLeaf 9
MaizeEarlyKharifHealthyLeaf 1
MaizeKharifCobFormationSulfurDeficiencyLeaf 1
MaizeKharifGrainFillingNoInsectMaydisLeafBlightDownyMildewNoDeficiency 6
MaizeEarlyKharifStemBorerLeaf 8
MaizeEarlyKharifSulfurDeficiencyLeaf 8
MaizeKharifEarlyVegetativeNoInsectDownyMildewSulphur 1
MaizeKharifGrainFillingStemBorerDownyMildewNoDeficiency 1
MaizeKharifGrainFillingNoInsectMaydisLeafBlightSulphur 1
MaizeKharifEarlyVegetativeStemBorerPolysoraRustNoDeficiency 1
MaizeEarlyKharifStemBorer,EyeSpotLeaf 1
MaizeKharifSeedlingStageStemBorerLeaf 1
MaizeEarlyKharifLeafSpot,EyeSpotLeaf 1
Maizeeyespot 1
MaizeEarlyKharifEyeSpot,LeafBlightLeaf 1
MaizeKharifEarlyVegetativeNoInsectFusariumStalkRotNoDeficiency 2
MaizeKharifGrainFilli

In [68]:
c =0
num = 0
for x in Dict:
    if "Maize" in x and "Insect" not in x:
        print(x, Dict[x])

MaizeKharifEarlyVegetativeStemBorerNoDiseaseNoDeficiency 228
MaizeKharifGrainFillingCobBorerNoDiseaseNoDeficiency 105
MaizeLeafspot 1
MaizeKharifGrainFillingStemBorerNoDiseaseNoDeficiency 77
MaizeKharifTilleringStemBorerNoDiseaseSulphur 6
MaizeEarlyKharifLeafSpotLeaf 3
MaizeEarlyKharifEyeSpot,StemBorerLeaf 3
MaizeKharifEarlyVegetativeStemBorerNoDiseaseSulphur 18
Maize 9
MaizeEarlyKharifLeafBlight,EyeSpotLeaf 1
MaizeKharifEarlyVegetativeCobBorerNoDiseaseNoDeficiency 14
MaizeEarlyKharifEyeSpotLeaf 9
MaizeEarlyKharifHealthyLeaf 1
MaizeKharifCobFormationSulfurDeficiencyLeaf 1
MaizeEarlyKharifStemBorerLeaf 8
MaizeEarlyKharifSulfurDeficiencyLeaf 8
MaizeKharifGrainFillingStemBorerNoDiseaseSulphur 10
MaizeKharifTilleringStemBorerNoDiseaseNoDeficiency 4
MaizeKharifGrainFillingStemBorerDownyMildewNoDeficiency 1
MaizeKharifEarlyVegetativeStemBorerPolysoraRustNoDeficiency 1
MaizeEarlyKharifStemBorer,EyeSpotLeaf 1
MaizeKharifEarlyMaturityCobBorerNoDiseaseNoDeficiency 1
MaizeKharifSeedlingStageStemB

In [101]:
c =0
num = 0
for x in Dict:
    if "Maize" in x and "Insect" not in x:
        print(x, Dict[x])

MaizeKharifEarlyVegetativeStemBorerNoDiseaseNoDeficiency 228
MaizeKharifGrainFillingCobBorerNoDiseaseNoDeficiency 105
MaizeLeafspot 1
MaizeKharifGrainFillingStemBorerNoDiseaseNoDeficiency 77
MaizeKharifTilleringStemBorerNoDiseaseSulphur 6
MaizeEarlyKharifLeafSpotLeaf 3
MaizeEarlyKharifEyeSpot,StemBorerLeaf 3
MaizeKharifEarlyVegetativeStemBorerNoDiseaseSulphur 18
Maize 9
MaizeEarlyKharifLeafBlight,EyeSpotLeaf 1
MaizeKharifEarlyVegetativeCobBorerNoDiseaseNoDeficiency 14
MaizeEarlyKharifEyeSpotLeaf 9
MaizeEarlyKharifHealthyLeaf 1
MaizeKharifCobFormationSulfurDeficiencyLeaf 1
MaizeEarlyKharifStemBorerLeaf 8
MaizeEarlyKharifSulfurDeficiencyLeaf 8
MaizeKharifGrainFillingStemBorerNoDiseaseSulphur 10
MaizeKharifTilleringStemBorerNoDiseaseNoDeficiency 4
MaizeKharifGrainFillingStemBorerDownyMildewNoDeficiency 1
MaizeKharifEarlyVegetativeStemBorerPolysoraRustNoDeficiency 1
MaizeEarlyKharifStemBorer,EyeSpotLeaf 1
MaizeKharifEarlyMaturityCobBorerNoDiseaseNoDeficiency 1
MaizeKharifSeedlingStageStemB

In [103]:
c =0
num = 0
for x in Dict:
    if "Borer" in x and "Maize" in x:
        c=c+1
        num = num + Dict[x]

print (num)

484


In [100]:
c =0
num = 0
for x in Dict:
    if "Maize" in x and "Potash" in x:
        num = num + Dict[x]
print(num)

2


In [39]:
c =0
num = 0
for x in Dict:
    if "NoDisease" in x and "Maize" in x:
        c=c+1
        num = num + Dict[x]

print (num)

696


In [44]:
c =0
num = 0
for x in Dict:
    if "DownyMildew" in x and "Maize" in x:
        c=c+1
        num = num + Dict[x]

print (num)

33


In [42]:

c =0
num = 0
for x in Dict:
    if "EyeSpot" in x and "Maize" in x:
        c=c+1
        num = num + Dict[x]

print (num)

16


## Red Gram

In [51]:
c =0
num = 0
for x in Dict:
    if "RedGram" in x and "Disease" not in x:
        print(x, Dict[x])

RedGramKharifEarlyVegetativeNoInsectSterilityMosaicNoDeficiency 7
RedGramKharifEarlyVegetativeNoInsectRedGramYellowMosaicNoDeficiency 9
RedGramKharifEarlyVegetativeNoInsectWiltNoDeficiency 4
RedGramKharifEarlyVegetativePodBugSterilityMosaicNoDeficiency 1
RedGramKharifEarlyVegetativeLeafFolder 1


In [66]:
c =0
num = 0
for x in Dict:
    if "RedGram" in x and "Deficiency" not in x:
        print(x, Dict[x])

RedGramKharifEarlyVegetativeLeafFolder 1


In [72]:
c =0
num = 0
for x in Dict:
    if "RedGram" in x and "Insect" not in x:
        print(x, Dict[x])

RedGramKharifEarlyVegetativeLeafHopperNoDiseaseNoDeficiency 32
RedGramLeafFolderKharifEarlyVegetativeLeafHopperNoDiseaseNoDeficiency 1
RedGramKharifEarlyVegetativeCowBugsNoDiseaseNoDeficiency 2
RedGramKharifEarlyVegetativeGramPodBorerNoDiseaseNoDeficiency 5
RedGramKharifEarlyVegetativePodBugSterilityMosaicNoDeficiency 1
RedGramKharifEarlyVegetativeLeafFolderNoDiseaseNoDeficiency 2
RedGramKharifEarlyVegetativeLeafFolder 1


In [107]:
c =0
num = 0
for x in Dict:
    if "CowBugs" in x and "RedGram" in x:
        c=c+1
        num = num + Dict[x]

print (num)

2


In [58]:
c =0
num = 0
for x in Dict:
    if "Mosaic" in x and "RedGram" in x:
        c=c+1
        num = num + Dict[x]

print (num)

17


# Deficiency

In [60]:
c =0
num = 0
for x in Dict:
    if "NoDeficiency" in x:
        c=c+1
        num = num + Dict[x]

print (num)

3373


## Insect

In [69]:
c =0
num = 0
for x in Dict:
    if "Insect" in x:
        c=c+1
        num = num + Dict[x]

print (num)

2266


In [70]:
c =0
num = 0
for x in Dict:
    if "Disease" in x:
        c=c+1
        num = num + Dict[x]

print (num)

3395


## Results of the analysis 

There are 3740 total images, 3395 have no diseases, 3382 have no deficiencies, 2266 have no insects <br>

The top crops present in the data are Cotton : 1181, Maize: 803, Red Gram : 506, Rice 350 and Black Gram: 136 <br>

Within Cotton: <br>
1745 images have no diseases <br>
121 have a leaf curl disease <br>
16 have a leaf spot disease <br>

1848 images have no deficiency <br>
24 have a magnesium related deficiency<br> 
3 have a boron deficiency and 5 have a potash deficiency <br>

1111 images have no insects <br>
455 leaf hoppers<br>
189 have mealy bugs<br>
67 white flies <br>
7 cut worms <br>


Within Maize:<br>

696 images have no diseases<br>
33 have midlew related diseases<br>
29 are blight related diseases<br>
16 eye spots<br>
5 Leaf spots<br>

654 have no deficiency<br>
102 have a sulpur deficiency<br>
3 have an iron deficiency<br>
2 had a potash deficiency<br>

283 have have no insects <br>
484 have a borer type insect<br>


Within Red Gram:<br>
<br>
544 have no disease<br>
17 have a mosaic disease<br>
4 have a wilt <br>
<br>
565 have no deficiency<br>
<br>
522 have no insects<br>
33 leaf hoppers<br>
2 cowbugs<br>


the reason why some of the values don't add up is because some of the labels are missing certain parts of their label which is why they cannot be classified<br>
for example RedGramKharifEarlyVegetativeLeafFolder<br>
where we don't know what the disease or deficiency is <br>

As for rice and black gram the folders get too small to allow for successful training of an algorithm 